pkg/controllers/management/auth/roletemplates/crtb_handler.go

```diff
@@ -0,0 +1,190 @@
+
+func newCRTBHandler(management *config.ManagementContext) *crtbHandler {
+	return &crtbHandler{
+		s:              status.NewStatus(),
+		userMGR:        management.UserManager,
+		userController: management.Wrangler.Mgmt.User(),
+		rtController:   management.Wrangler.Mgmt.RoleTemplate(),
+		crController:   management.Wrangler.RBAC.ClusterRole(),
+		crbController:  management.Wrangler.RBAC.ClusterRoleBinding(),
+	}
+}
+
+// OnChange syncs the required resources used to implement a CRTB. It does the following:
+//   - Create the specified user if it doesn't already exist.
+//   - Create the membership bindings to give access to the cluster.
+//   - Create a binding to the project management role if it exists.
+//   - Create a binding to the cluster management role if it exists.
+//
+// TODO add statuses
+func (c *crtbHandler) OnChange(_ string, crtb *v3.ClusterRoleTemplateBinding) (*v3.ClusterRoleTemplateBinding, error) {
+	var localConditions []metav1.Condition
+
+	// Create user
+	crtb, err := c.reconcileSubject(crtb, &localConditions)
+	if err != nil {
+		return nil, err
+	}
+
+	rt, err := c.rtController.Get(crtb.RoleTemplateName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create membership binding
+	if err := createOrUpdateMembershipBinding(crtb, rt, c.crbController); err != nil {
+		return nil, err
+	}
+
+	crbsNeeded := map[*rbacv1.ClusterRoleBinding]bool{}
+	ownerLabel := rbac.CreateCRTBOwnerLabel(crtb.Name)
+
+	// Check if there is a project management role to bind to
+	projectMagementRoleName := rbac.ProjectManagementPlaneClusterRoleNameFor(crtb.RoleTemplateName)
+	cr, err := c.crController.Get(projectMagementRoleName, metav1.GetOptions{})
+	if err == nil && cr != nil {
+		crb, err := rbac.BuildClusterRoleBindingFromRTB(crtb, ownerLabel, projectMagementRoleName)
+		if err != nil {
+			return nil, err
+		}
+		crbsNeeded[crb] = true
+	} else if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Check if there is a cluster management role to bind to
+	clusterManagementRoleName := rbac.ClusterManagementPlaneClusterRoleNameFor(crtb.RoleTemplateName)
+	cr, err = c.crController.Get(clusterManagementRoleName, metav1.GetOptions{})
+	if err != nil && cr != nil {
+		crb, err := rbac.BuildClusterRoleBindingFromRTB(crtb, ownerLabel, clusterManagementRoleName)
+		if err != nil {
+			return nil, err
+		}
+		crbsNeeded[crb] = true
+	} else if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	currentCRBs, err := c.crbController.List(metav1.ListOptions{LabelSelector: ownerLabel})
+	if err != nil {
+		return nil, err
+	}
+
+	// Find if CRBs already exists
+currentLoop:
+	for _, currentCRB := range currentCRBs.Items {
+		for desiredCRB := range crbsNeeded {
+			if rbac.AreClusterRoleBindingsSame(&currentCRB, desiredCRB) {
+				crbsNeeded[desiredCRB] = false
+				continue currentLoop
+			}
+			if err := c.crbController.Delete(currentCRB.Name, &metav1.DeleteOptions{}); err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
I'm confused by this logic (as well as concerned about the nested for loop).
Also, wouldn't it really only iterate over the first item in `crbsNeeded`?:
 - If `AreClusterRoleBindingsSame` is `true`, it `continue`s the outer loop
 - If `false`, it directly deletes `currentCRB`, but it will continue the inner loop despite being already deleted?
Please take another look since I believe there may be something wrong (or worth explaining otherwise)

Reply:
I think I wrote this at the end of my day and definitely over-complicated things. I think my new version is much clearer. Using labels/gotos should have been my sign to come back another day 🤣